ztools/ztls/handshake_client.go

```diff
@@ -32,19 +32,235 @@ type clientHandshakeState struct {
+	CacheKey     CacheKeyGenerator
+}
+
+type ClientExtension interface {
+	Marshal() []byte
+	CheckImplemented() error
+}
+
+func (c *ClientFingerprintConfiguration) CheckImplementedExtensions() error {
+	for _, ext := range c.Extensions {
+		if err := ext.CheckImplemented(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *ClientFingerprintConfiguration) ModifyConfig(config *Config) *Config {
+	config.NextProtos = []string{}
+	config.CipherSuites = c.CipherSuites
+	config.MaxVersion = c.HandshakeVersion
+	config.ClientRandom = c.ClientRandom
+	config.CurvePreferences = []CurveID{}
+	config.HeartbeatEnabled = false
+	config.ExtendedRandom = false
+	config.ForceSessionTicketExt = false
+	config.ExtendedMasterSecret = false
+	config.SignedCertificateTimestampExt = false
+	for i, ext := range c.Extensions {
+		switch casted := ext.(type) {
+		case SniExtension:
+			if casted.Autopopulate {
+				c.Extensions[i] = SniExtension{[]string{config.ServerName}, true}
+			}
+			if config.ServerName == "" && len(casted.Domains) > 0 {
+				config.ServerName = casted.Domains[0]
+			}
+		case ALPNExtension:
+			config.NextProtos = casted.Protocols
+		case ExtendedMasterSecretExtension:
+			config.ExtendedMasterSecret = true
+		case SignatureAlgorithmExtension:
+			supportedSKXSignatureAlgorithms = casted.getStructuredAlgorithms()
+			defaultSKXSignatureAlgorithms = casted.getStructuredAlgorithms()
+		case SCTExtension:
+			config.SignedCertificateTimestampExt = true
+		case SupportedCurvesExtension:
+			config.CurvePreferences = casted.Curves
+		case SessionTicketExtension:
+			config.ForceSessionTicketExt = true
+		default:
+			continue
+		}
+	}
+	return config
+}
+
+func (c *ClientFingerprintConfiguration) marshal(config *Config) ([]byte, error) {
+	if err := c.CheckImplementedExtensions(); err != nil {
+		return nil, err
+	}
+	head := make([]byte, 38)
+	head[0] = 1
+	head[4] = uint8(c.HandshakeVersion >> 8)
+	head[5] = uint8(c.HandshakeVersion)
+	if len(c.ClientRandom) == 32 {
+		copy(head[6:38], c.ClientRandom[0:32])
+	} else {
+		_, err := io.ReadFull(config.rand(), head[6:38])
+		if err != nil {
+			return nil, errors.New("tls: short read from Rand: " + err.Error())
+		}
+	}
+
+	sessionID := make([]byte, len(c.SessionID)+1)
+	sessionID[0] = uint8(len(c.SessionID))
+	if len(c.SessionID) > 0 {
+		copy(sessionID[1:], c.SessionID)
+	}
+
+	ciphers := make([]byte, 2+2*len(c.CipherSuites))
+	ciphers[0] = uint8(len(c.CipherSuites) >> 7)
+	ciphers[1] = uint8(len(c.CipherSuites) << 1)
+	for i, suite := range c.CipherSuites {
+		if !config.ForceSuites {
+			found := false
+			for _, impl := range implementedCipherSuites {
+				if impl.id == suite {
+					found = true
+				}
+			}
+			if !found {
+				return nil, errors.New(fmt.Sprintf("tls: unimplemented cipher suite %d", suite))
+			}
+		}
+
+		ciphers[2+i*2] = uint8(suite >> 8)
+		ciphers[3+i*2] = uint8(suite)
+	}
+
+	compressions := make([]byte, len(c.CompressionMethods)+1)
```

Comment:
Shouldn't be a problem with the checks below, but adding because safety is best.